### PR TITLE
Implement dump/load VmContext

### PR DESCRIPTION
**Pull request by Crazykev**
This is runv part of hyperd save/restore.

### CODE REVIEW

**Comment by Crazykev** [comment]
@laijs ready for review

File: hyperstart/libhyperstart/persistence.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+		LastStreamSeq: h.lastStreamSeq,
+		Procs:         make(map[string]persistProcessState, len(h.procs)),
+	}
+	for k, v := range h.procs {
+		p.Procs[fmt.Sprintf("%s@%s", k.c, k.p)] = persistProcessState{
+			StdioSeq:  v.stdioSeq,
+			StderrSeq: v.stderrSeq,
+			StreamSeq: v.stdinPipe.streamSeq,
+		}
+	}
```
Comment (laijs) [inline hyperstart/libhyperstart/persistence.go]: vmContext has all the info(xxxxSeq) of the containers, please use the info from vmContext.

and 'Dump()' is unneeded, libhyperstart has already provide LastStreamSeq(), please use that it.

File: hyperstart/libhyperstart/serial.go [hunk 2]
```diff
@@ -645,6 +649,21 @@ func (h *serialBasedHyperstart) NewContainer(c *hyperstartapi.Container) (io.Wri
 	return &ps.stdinPipe, ps.stdoutPipe, ps.stderrPipe, err
 }
 
+func (h *serialBasedHyperstart) RestoreContainer(c *hyperstartapi.Container) (io.WriteCloser, io.ReadCloser, io.ReadCloser, error) {
+	ps, ok := h.procs[pKey{c: c.Id, p: c.Process.Id}]
```
Comment (laijs) [inline hyperstart/libhyperstart/serial.go]: please create pState in this function. "load()" is unneeded.

File: hypervisor/context.go [hunk 3]
```diff
@@ -13,9 +13,8 @@ import (
 )
 
 type VmHwStatus struct {
-	PciAddr  int    //next available pci addr for pci hotplug
-	ScsiId   int    //next available scsi id for scsi hotplug
-	AttachId uint64 //next available attachId for attached tty
+	PciAddr int //next available pci addr for pci hotplug
+	ScsiId  int //next available scsi id for scsi hotplug
```
Comment (laijs) [inline hypervisor/context.go]: AttachId is still needed for compatiblity, save LastStreamSeq() to here.

**Comment by laijs** [comment]
compatiblity is the most important thing to be consider. the new code should compatible with the old saved data. and daemon can be restored with the old saved data.

File: hypervisor/persistence.go [hunk 4]
```diff
@@ -19,20 +30,29 @@ type PersistVolumeInfo struct {
...
+	LoList   []*PersistInterfaceInfo
+}
+
 type PersistInfo struct {
-	Id          string
-	DriverInfo  map[string]interface{}
-	VmSpec      *hyperstartapi.Pod
-	HwStat      *VmHwStatus
-	VolumeList  []*PersistVolumeInfo
-	NetworkList []*PersistNetworkInfo
```
Comment (laijs) [inline hypervisor/persistence.go]: PersistVersion int  // save it save 20170223

File: hypervisor/persistence.go [hunk 5]
```diff
@@ -145,46 +242,54 @@ func (pinfo *PersistInfo) vmContext(hub chan VmEvent, client chan *types.VmRespo
...
+			DiskDescriptor: binfo,
+			sandbox:        ctx,
+			observers:      make(map[string]*sync.WaitGroup),
+			lock:           &sync.RWMutex{},
+			// FIXME: is there any trouble if we set it as ready when restoring from persistence
+			ready: true,
+		}
+	}
+
+	for _, pc := range pinfo.ContainerList {
```
Comment (laijs) [inline hypervisor/persistence.go]: you might need to synthesize pinfo.VolumeList & pinfo.ContainerList from pinfo.VmSpec if it is being restored from the old saved data

**Comment by Crazykev** [comment]
@laijs All checked, PTAL

**Pull request merged**